### PR TITLE
add wildcard parameter to allow resolving multiple name to the same IP

### PR DESCRIPTION
**Pull request by jremond**
See discussion https://github.com/coredns/coredns/issues/691

Implementation of a separate wildcard keyword to allow to resolve
any.thing.ip-1.2.3.4.example.org to 1.2.3.4

### CODE REVIEW

**Comment by miekg** [comment]
@pmoroney Want to give this one a try?
(I will also give it a check)

**Comment by pmoroney** [comment]
Sure, I'll look into it in a few

File: middleware/reverse/setup.go [hunk 1]
```diff
@@ -117,8 +121,12 @@ func reverseParse(c *caddy.Controller) (nets networks, fall bool, err error) {
 				if ipnet.IP.To4() == nil {
 					regexIP = regexMatchV6
 				}
+				prefix := "^"
+				if wildcard {
+					prefix += ".*"
```
Comment (pmoroney) [inline middleware/reverse/setup.go:126]: nit - Removing the explicit ^ should work without needing to specify the .* at the beginning, so we can just set prefix to the empty string.

**Review by pmoroney — COMMENTED** [review]
Question: assuming the following template: {ip}.{zone}
do we want wildcard to return a result for test1.2.3.4.example.net or do we want to require a "." between the ip and the prefix if the template starts with "{ip}"? I am only asking because I am not sure about all the use cases for this middleware. 

Also, I would like to see a test of the wildcard functionality added.

**Comment by jremond** [comment]
Do you have an example of test (in another middleware) that you think would make sense ?

File: middleware/reverse/setup.go (hunk 1 above)
Comment (jremond) [reply middleware/reverse/setup.go:126]: After testing, setting prefix to "" does not work.

**Comment by miekg** [comment]
This will either make 009 or be pushed to 010 (I want to release 009 quickly because of bugs found)

**Comment by miekg** [comment]
Tests can be found in basically any middleware `TestChaos` in middleware/chaos is I think simplest.

File: middleware/reverse/setup.go (hunk 1 above)
Comment (miekg) [reply middleware/reverse/setup.go:126]: That doesn't make sense. 

~~~
diff --git a/middleware/reverse/setup.go b/middleware/reverse/setup.go
index ee426f85..663ebea6 100644
--- a/middleware/reverse/setup.go
+++ b/middleware/reverse/setup.go
@@ -121,7 +121,7 @@ func reverseParse(c *caddy.Controller) (nets networks, fall bool, err error) {
                                if ipnet.IP.To4() == nil {
                                        regexIP = regexMatchV6
                                }
-                               prefix := "^"
+                               prefix := ""
                                if wildcard {
                                        prefix += ".*"
                                }
diff --git a/middleware/reverse/setup_test.go b/middleware/reverse/setup_test.go
index 2198c1e3..65671a72 100644
--- a/middleware/reverse/setup_test.go
+++ b/middleware/reverse/setup_test.go
@@ -14,10 +14,10 @@ func TestSetupParse(t *testing.T) {
        _, net4, _ := net.ParseCIDR("10.1.1.0/24")
        _, net6, _ := net.ParseCIDR("fd01::/64")
 
-       regexIP6, _ := regexp.Compile("^ip-" + regexMatchV6 + "\\.domain\\.com\\.$")
-       regexIpv4dynamic, _ := regexp.Compile("^dynamic-" + regexMatchV4 + "-intern\\.dynamic\\.domain\\.com\\.$")
-       regexIpv6dynamic, _ := regexp.Compile("^dynamic-" + regexMatchV6 + "-intern\\.dynamic\\.domain\\.com\\.$")
-       regexIpv4vpndynamic, _ := regexp.Compile("^dynamic-" + regexMatchV4 + "-vpn\\.dynamic\\.domain\\.com\\.$")
+       regexIP6, _ := regexp.Compile("ip-" + regexMatchV6 + "\\.domain\\.com\\.$")
+       regexIpv4dynamic, _ := regexp.Compile("dynamic-" + regexMatchV4 + "-intern\\.dynamic\\.domain\\.com\\.$")
+       regexIpv6dynamic, _ := regexp.Compile("dynamic-" + regexMatchV6 + "-intern\\.dynamic\\.domain\\.com\\.$")
+       regexIpv4vpndynamic, _ := regexp.Compile("dynamic-" + regexMatchV4 + "-vpn\\.dynamic\\.domain\\.com\\.$")
 
        serverBlockKeys := []string{"domain.com.:8053", "dynamic.domain.com.:8053"}
 
~~~
fixes the tests

Comment (jremond) [reply middleware/reverse/setup.go:126]: I don't mean the tests are broken, I mean 
test.ip-127.0.3.1.example.org does not resolve to 127.0.3.1 if I set the prefix to ""
I get a SERVFAIL in the logs

**Comment by miekg** [comment]
Ah. That's a different thing then.

**Comment by jremond** [comment]
@pmoroney I just added a test based on the one from the chaos middleware, let me know if it is what you asked for.
I'm still unable to make it work with an empty prefix instead of ".*", any hint on that ?

**Comment by pmoroney** [comment]
The new test in reverse_test.go looks good, thank's for adding that. 
On line 126 of setup.go, if I change it to: `prefix = ""` it seems to work for me.

And also, while I was trying to figure out why the patch code coverage target wasn't hit, I realized that the added test in setup_test.go might not actually be testing anything. It might just work to copy the first test in the table (the `// with defaults` one), add the wildcard config flag, and make sure the regexp changed appropriately. Hopefully then the check will pass :)

**Comment by jremond** [comment]
my Corefile is 
```
.:1053 {
    log stdout
    errors stdout
    reverse 127.0.0.0/8 {
        hostname ip-{ip}.example.org
        wildcard
    }
}
```
if I change the line 126 of setup.go to `prefix = ""`
then `dig +short  @127.0.0.1 -p 1053 test.ip-127.0.3.1.example.org`
does not return anything and the corresponding log line is 
```
.:1053
2017/07/10 21:57:09 [INFO] CoreDNS-008
CoreDNS-008
127.0.0.1 - [10/Jul/2017:21:57:21 -0400] "A IN test.ip-127.0.3.1.example.org. udp 59 false 4096" SERVFAIL 59 180.182µs
```

Can you share your `Corefile` ?


I pushed a new test in setup_test.go to see if I understand correctly what is expected.

**Comment by pmoroney** [comment]
I was using the following corefile:
```
example.org:5533 {
    reverse 10.0.0.0/8 {
        wildcard
    }
}
```
And I was getting the following response:
```
$ dig @127.0.0.1 -p 5533 asip-10.1.2.3.example.org                                                                                                                        

; <<>> DiG 9.10.3-P4 <<>> @127.0.0.1 -p 5533 asip-10.1.2.3.example.org
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 2423
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;asip-10.1.2.3.example.org.     IN      A

;; ANSWER SECTION:
asip-10.1.2.3.example.org. 60   IN      A       10.1.2.3

;; Query time: 0 msec
;; SERVER: 127.0.0.1#5533(127.0.0.1)
;; WHEN: Sat Jul 08 23:50:16 MDT 2017
;; MSG SIZE  rcvd: 70
```

But it works either way, the new test looks great, and all the checks pass. So I am going to merge it anyways. Thanks again!

**Pull request merged**